### ISSUE TITLE
Standards for language level (and maybe coding style)

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1158]
It would be nice to set down standards for what code can be contributed to igraph.

 - What language versions are allowed?  C89 / C99?  C++98/11/14/17?

 - I noticed that @gaborcsardi made several changes to comply with CRAN's rules. It may be useful to impose some of the more reasonable rules directly on igraph.

----

What I am looking for personally is to start allowing C++11. Not a lot new code is being written in C++98 anymore and sticking to this old version will limit what libraries can be integrated into igraph.  I am not sure if the requirements are currently documented anywhere, but igraph does not contain any C++11 code at this moment.

As for C, I'd suggest sticking with C89 because MSVC (an important compiler) does not support C99 properly.

**Comment by szhorvat** [comment]
One potential counterargument for allowing C++11 is that RHEL 6 is still being supported ("maintenance support") until 2020, and it comes with gcc 4.4 as the default compiler (i.e. no C++11 support). Several distros are based on RHEL, such as CentOS or Scientific Linux.

Some software, such as Mathematica, still use RHEL 6 as a compatibility baseline.

RHEL 7 already comes with gcc 4.8, which does support it.

**Comment by ntamas** [comment]
I'd stick to C89 on the C side and C++98 on the C++ side for the reasons mentioned above, at least until RHEL reaches EOL.

**Comment by szhorvat** [comment]
It seems RHEL6 is already in the "Maintenance Support 2" phase, which is rather limited support: 

https://access.redhat.com/support/policy/updates/errata

C++11 would allow us to integrate vf3lib, the successor of VF2. It claims to be much faster (also much faster than LAD).

https://github.com/vincarlet/vf3lib

**Comment by szhorvat** [comment]
Speed comparison from the paper:

![image](https://user-images.githubusercontent.com/1212871/56462263-9a977a80-63c0-11e9-8e1f-999e33caafaa.png)

For plain isomorphism, we already have the excellent Bliss of course. What this would provide over that is subgraph matching and matching with coloured edges.

**Comment by szhorvat** [comment]
Well, vf3lib seems to compile just fine as C++98 despite the claims that it's C++11 ...

**Comment by iosonofabio** [comment]
Guys, Do you think this is really necessary for 0.8? I would slip it into 0.9+ since it seems somewhat inconsequential for now and a slog

**Comment by vtraag** [comment]
I wanted to re-open this discuss, not only regarding the language level (i.e. which standards of C and which standards of C++ to support), but also to open the question why we stick to having `igraph` as a C library, instead of taking advantage of some of the C++ features. Of course, the original suggestion by @szhorvat was not to switch completely to C++ of course, but I wanted to broaden this discussion and consider this.

Using C++ instead of C would have a number of benefits:

1. Allow operator overloading, i.e. using `[]` instead of having to use the `VECTOR` macro.
2. Allow namespaces (c.f. https://github.com/igraph/igraph/issues/1001)
3. Allow function name overloading (i.e. allow simpler call signatures).
4. Allow STL instead of maintaining our own implementations (c.f. [discussion on data types](https://igraph.discourse.group/t/data-structure-type-usage-and-visibility/316/2)).
5. Allowing templating, so we no longer need to depend on the poor-man's template (c.f. https://github.com/igraph/igraph/pull/1429) for types that are not included in the STL.
6. Allowing exception handling instead of maintaining our own implementation using `FINAL`.
7. Allow the use of streams, instead of having to deal with `FILE` pointers.
8. Easier iterating over variables, improving the readability.

Overall, I think the readability of `igraph` code would greatly improve. This would also make it easier for external contributors to start contributing. Because of the reduced need to implement some of our own data types, I think it will also reduce the maintenance costs. There are probably some additional advantages as well, feel free to add.

As far as I understand, we can compile for all still supported Python versions (i.e. Python 3.5 and up) using VS2019 (VC 14.2), which [fully supports the (core) C++11 standard](https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=vs-2019). RHEL 6 is indeed only on maintenance support, and additionally it will [end in November this year](https://access.redhat.com/support/policy/updates/errata/) anyway. We could therefore switch to C++11 fully I think.

What are the reasons for keeping `igraph` a C library? Of course, it would no longer make the `igraph` library available for coding in C, but most people are using the higher-level interfaces anyway. Looking forward to hearing your thoughts on this. Of course, switching to C++11 would require a considerable time investment, but I think it would reduce maintenance costs in the future.

**Comment by szhorvat** [comment]
It would require more thought to write a proper response, so for now I would just like to share a few random ideas that sprang to mind, without explicitly arguing for any specific decision.

I understand where you are coming from and I share your frustration. Once one has tasted the convenient of C++, igraph does not seem that nice to work with. Despite that, at this point I am still quite sceptical about the possibility to move to C++ ...

 - There is value in having a C API. C++ ABIs are very complex, and as a consequence tend to change quickly and be incompatible. On Windows, generally, it is possible to have a C API DLL that works with just about anything. There are caveats (hence the recommendations to use a certain MSVC version with a certain Python version) but generally things work. With C++, compatibility goes out the window. 

 - igraph is designed to be embeddable into various high-level languages. Basically every language has a C foreign function interface. Very few have one for C++.

 - Note that the above two points refer to the API, not the internal implementation. Because of the great advantages that come with having a C API, even many libraries that are written in C++ have an additional plain C interface.

 - igraph *could*, in principle, have a an additional C++ interface. I believe @ntamas started one at some point? If it had one, there is no reason why this interface could not be a bit better integrated and also used internally. This could relieve *some* of the pain during igraph development, but not all (e.g. it wouldn't make FINALLY go away). It would make igraph friendlier and easier to use for external users (not internal development).

 - The changes you propose are extensive. I fear that getting there would take a huge amount of work, perhaps even comparable to starting from scratch. I mentioned at the beginning of this post that I do not love programming igraph. Why am I here then? Because I believe that it's better to build on an existing foundation than starting from scratch.

 - If igraph were to transition to C++, it would be viewed very differently. It would compete with libraries such as Boost Graph and LEMON which make extensive use of advanced C++ features such as templates. Actually, from my perspective, *not* making use of these features, and keeping things simple and predictable, can be an advantage. Many scientists (not software engineers!) actually prefer C because it's easier to deal with than the complex beast that C++ is. This is *not* an argument against C++. It's an argument for keeping things "C-like" even if we were to go with C++.

Overall, the reason why I am so sceptical is that this feels like a **huge** amount of work, bordering on a complete rewrite. Doing that is always tempting, but in my opinion usually not a good idea. It is very easy to underestimate the amount of work and overestimate the gains (the end result is never perfect).

Also, there are good reasons to keep a C API, regardless of how things are implemented internally.

An middle way could be:

 - keep the C API as the primary API
 - add an official C++ interface, which will make using igraph much more pleasant
 - start using this C++ interface internally to speed up development

I can't tell at this point if this would work out at all, and by how much it would speed up future development. We would need to examine every pain point you listed, and see if it helps with them. It might not work at all. E.g., throwing exceptions through C functions is not guaranteed to work, which means that exception use would be limited unless we go all-C++. If we have to stick with `FINALLY`, then I question the value of partially moving to C++.

**Comment by szhorvat** [comment]
On another note, at this point I would very much like to start allowing selected C99 features that are well-supported by MSVC, as well as C++11. If we make `igraph_integer` 64-bit, it will be easier to use an `int64_t` than have to detect the size of `short`, `int`, `long` (which can be just about anything, not just in theory, [but also in practice](https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models)).

**Comment by iosonofabio** [comment]
@vtraag Vacation did you some good, I love the disruptive attitude!

What you say is so true. Then if you have C++11 you can do Python with pybind11 instead of suffering through the manual C API. So clean, much more maintainable. That is why all my repos that need speed are coded that way (or variations such as SWIG).

But hey, @szhorvat is right, you're not talking about igraph, what you are describing is neograph, a yet-to-be-born library that everyone will use in 2035.

Before landing on igraph, I used graph-tool for a bit. Man, that guy is impressive, almost belongs to another era of coding. Super efficient C++ code with small codebase thanks to heavy templating metaprogramming, and an equally elegant Python verandah. But you know what, I've been a power user of Arch Linux for like 15 years and it took me days to compile it. It takes so long that I thought there must be a config bug in the AUR package. Turns out that is just the way it is, you have to wait a day because the compiler (GCC) is not advanced enough to digest the elegance. The dev told me that will improve with future compilers, and I am wondering whether that argument works with my funding agency 😋

Then I tried to add a function and it all fell apart. After days I still couldn't for the best of me figure out where to add what code. It felt like a jewelry, if you touch anything it's going to ruin everything. Plus, remember you can't really compile that often to check for bugs, because it takes a day to do so.

Bottom line: I am happy to slowly switch and python won't be a problem, but let's make sure:
- it is gradual
- we stop before it's too late (or too elegant)

**Comment by vtraag** [comment]
To be sure, my proposal would not be to *add* an additional C++ interface. That would only increase the maintenance costs, instead of decreasing the maintenance costs. My proposal would be to have C++ internally, possibly with an additional C interface so that it is usable from a C library. Or, as @iosonofabio suggests, use tools such as `pybind11` or `SWIG` to generate interfaces for higher-level languages.

As for comparing `igraph` with other graph libraries, I think Boost graph and LEMON have a different target audience. Compared to `igraph`, the number of available algorithms/functions seem more limited. Probably `graph-tool` comes closer to what `igraph` has to offer, but does not offer support for Windows, and is only available in Python (perhaps it can also be used as a stand-alone library, I am not sure). I haven't looked in detail at the source code of `graph-tool`, I know it relies on templating extensively, and that it takes a while to compile. It is nice to hear you think the approach Peixoto has taken is very elegant @iosonofabio! But if it indeed prevents others from contributing, I would think that might make it difficult to support the package in the long term.

The most important consideration I had for proposing to use C++ is that I want to make it as easy as possible to read the code, and hence, contribute code to `igraph`. Maintaining things that rely on a lot of custom code would require one to first learn all of that custom code before one can start to contribute or to fix bugs. If, instead, much of the code is familiar standard C++, it is much easier to understand the code, and additionally, much easier to contribute or fix something. In short, my aim is to make the code as easy to read and as easy to understand as possible for everyone. This also means that the "elegance of code" should not be the focus, but the "readability of code" should be the focus. 

Of course, we could also remove some of the code if we rely on C++ (e.g. vectors and some other data types), which also reduces the maintenance costs.

I understand your concern that it would take a lot of time to make this transition completely. Clearly, this is also the biggest concern. We all agree on some of the benefits of using C++ I think, but it is not clear whether the benefits outweigh the costs. If we agree that it might be useful for us to start allowing C++ in the core of `igraph` (besides the external contributions which are already in C++ anyway), perhaps we could identify areas in which these changes would have the largest benefit.

What areas would you think would bring the greatest benefit with the least costs?

**Comment by iosonofabio** [comment]
@szhorvat @gaborcsardi are Mathematica and R compatible with C++? How messy would it be for those lands?

**Comment by vtraag** [comment]
Regarding R, `SWIG` can generate an interface from C/C++, see http://swig.org/Doc1.3/R.html.

**Comment by iosonofabio** [comment]
Yes, and also SWIG has reached version 4 ;-)

But SWIGging is a whole dark art, e.g. my previous crazyness:

https://github.com/neherlab/ffpopsim/blob/master/src/python/ffpopsim_highd.i

so let's hear from others too

**Comment by ntamas** [comment]
I generally agree with the opinion of @szhorvat, especially on the points that we probably genuinely underestimate how hard and time consuming it would be to switch to C++ and not break all the things along the road. My guesstimate is that it would well exceed the scope / budget of the CZI funding :)

As for the C++ interface to igraph, [this repo](https://github.com/ntamas/igraphpp) is the one that @szhorvat was talking about. I started it a long time ago for [another project of mine](https://github.com/ntamas/blockmodel) where I needed a decent C++ interface for igraph, and I only implemented the parts that I actually needed, hoping that I would notice common patterns along the way and then I could simply generate the rest of the interface somehow. I think that `igraphpp` is still pretty decent and probably wouldn't need much work to get it finished up to an extent that it would actually allow people to work with igraph from C++ -- but it still wouldn't simplify the internals of the C core.

The example of [`graph-tool`](https://graph-tool.skewed.de/) illustrates nicely what happens when you try to interface with a C++-based library from a higher level language. As far as I know, `graph-tool` is based on the Boost Graph Library, which is heavily based on templates. The net consequence is that it is somewhat faster (not sure whether we could not squeeze more performance from igraph, though, if we tried), it has nice in-place graph filtering capabilities (made possible by the templating magic), but it takes substantially more resources to compile (it does not matter for end users in higher langages, though, as long as the higher level language provides support for pre-compiled extensions) and it cannot support Windows sensibly. The latter is probably due to the heavy use of templates -- if we restrain ourselves and limit the usage of templates to only the parts that the STL supports anyway, we might get around with it and still keep Windows compatibility. But it nevertheless does not change the fact that it would be an enormous amount of work to switch igraph to C++ to a point where we really start seeing the benefits of it, and there would be a long (years-long) transition period where we would have to deal with a strange half-C-half-C++ hybrid.

**Comment by szhorvat** [comment]
Noting that now we require from the compiler:

 - C++11 support
 - Partial C99 support, in particular C99 standard library support.

Suggested baseline is VS 2015 (i.e. compiler version 19.0). Links to docs are here: https://stackoverflow.com/a/48615212/695132  Note that VS focuses on C++: it has excellent C++ standards conformance, but tends to only support those C features which are required by C++ (hence good C99 standard library support but incomplete language support).

If you need to test if a certain feature is supported by MSVC, https://godbolt.org/ is convenient. Choose "x64 msvc v19.0 (WINE)" and be sure to set the language to "C" if testing C features.

**Comment by szhorvat** [comment]
Added a note about language standards to https://github.com/igraph/igraph/wiki/Tips-on-writing-igraph-code  This resolves the problem in the original post above, so closing this (the discussion got a bit derailed). Re-open if you disagree.
